Fix screen-space aa methods when using nearest-neighbour scaling

Kaleb-Reid opened the pull request:
This fixes the issue brought up by https://github.com/godotengine/godot/pull/79731#discussion_r2565719927.
||Before|After|
|:-------------------------:|:-------------------------:|:-------------------------:|
|**None**|<img width="1367" height="839" alt="image" src="https://github.com/user-attachments/assets/854acc4d-76c4-46ef-a5bb-6dc31e4806b9" />||
|**FXAA**|<img width="1366" height="839" alt="image" src="https://github.com/user-attachments/assets/7cfd125d-1c2c-4849-9fdc-8c2816a67668" />|<img width="1368" height="843" alt="image" src="https://github.com/user-attachments/assets/755de8a5-1965-4fb1-939e-426a32801b8f" />|
|**SMAA**|<img width="1366" height="839" alt="image" src="https://github.com/user-attachments/assets/901cfd02-16a3-4545-a692-0629b08f4129" />|<img width="1369" height="842" alt="image" src="https://github.com/user-attachments/assets/17b57b49-005f-4f33-b81a-0b63fb59168d" />|

**All images using 0.25 scale**

Kaleb-Reid committed:
Use correct filter for SMAA colour texture

danbolt commented:
Really cool to see!

I’m not sure what Allen/Calinou are thinking here but it might be good to include a before/after picture to show that it addressed the bug.

Kaleb-Reid commented:
@danbolt I've added screenshots to the pr

allenwp commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
This comment needs to be updated (in this PR or in the original PR).

allenwp commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -667,6 +667,9 @@ void RendererSceneRenderRD::_render_buffers_post_process_and_tonemap(const Rende
 	}
 
+	// Make the hardware perform nearest-neighbor filtering when scaling the viewport 3D buffer.
+	RS::CanvasItemTextureFilter filter_mode = rb->get_scaling_3d_mode() == RS::VIEWPORT_SCALING_3D_MODE_NEAREST ? RS::CANVAS_ITEM_TEXTURE_FILTER_NEAREST : RS::CANVAS_ITEM_TEXTURE_FILTER_LINEAR;
```
Compared to the original PR, this PR changes the filter mode used by the tonemap functions to nearest when spatial upscale and smaa are not enabled.

I don't understand this PR well enough to know what the correct approach is, but this configuration was working fine, so I'm not sure why it needs changing...

Kaleb-Reid replied:
@allenwp In terms of spatial upscaling, I'm pretty sure if the scaling mode is nearest that means that spatial upscaling is disabled (fsr / metal fx spatial), so this shouldn't make a difference.

The original PR is from before SMAA being merged and so the `use_smaa` condition for the branch was added recently and is part of why nearest currently breaks when enabling smaa.

allenwp replied:
Oooooh, the history helps this make sense. Thanks!

Calinou commented:
Judging by the before/after screenshots, this doesn't really seem to work correctly. The visual result should be 100% identical to running SMAA at a low-but-native window size, then resizing a screenshot in an image editor with nearest-neighbor filtering. FXAA already seems to be doing this correctly, owing to its single-pass nature.

You can try this out in e.g. [Tesseract](http://tesseract.gg/) with the `/smaa 1; gscale 25; gscalenearest 1` command in chat for what I would expect to happen.

No AA | FXAA | SMAA
-|-|-
![2025-11-27_22 54 00_complex_edit](https://github.com/user-attachments/assets/2e4e79f9-9434-4fc4-9e52-00d163b44a7a) | ![2025-11-27_22 54 09_complex_edit](https://github.com/user-attachments/assets/2542b164-2696-4dcd-9301-7c681c80d1a0) | ![2025-11-27_22 54 05_complex_edit](https://github.com/user-attachments/assets/aa6dad6e-4ee3-452e-b597-277b86fbd875)

Instead, what happens in this PR is that the antialiasing effect is nearly lost, and some pixelated lines become smooth (and detailed) instead:

<img width="142" height="70" alt="image" src="https://github.com/user-attachments/assets/b25f4249-6930-47b9-b69a-4935ffc1ec6e" />

This breaks the pixel art aesthetic that nearest-neighbor filtering goes for.

Kaleb-Reid commented:
@Calinou I think I found the problem so I just pushed the fix and updated the screenshot.
EDIT: Also now that I look at it FXAA has the same problem...

Calinou commented:
Seems to work now, thanks! However, FXAA still exhibits the issue this PR originally had (in both Forward+ and Mobile).

No AA | FXAA | SMAA
-|-|-
<img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/9c298ba0-8509-4085-befb-cabd75960ddd" /> | <img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/b78d450b-8e8e-4357-ac3f-84fbe1384db0" /> | <img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/a73a1e6f-180b-4256-9537-88b685e0fbfb" />

Could you see if a similar fix can be applied to FXAA? If applied correctly, it should look similar to SMAA, except a little blurrier and with worse edge coverage.

Kaleb-Reid commented:
@Calinou ~I've modified it to preserve pixel sizes when using FXAA as well but there is some other issue which basically causes FXAA to do nothing when applying nearest scaling. You can easily see this if you compare 0.9999 3d scale and 1.0, but it is still doing _something_ because if you go to 0.1 scale and turn FXAA on / off the aliasing appears worse as thin objects have new gaps in them.~

I've fixed FXAA now, the changes in the original pr to `tonemapper()` were causing the remaining issue. I've reverted those changes and it appears to perform FXAA properly. Something someone should look at is if my recent changes break multiview, I'm not sure how to test it.

Calinou commented:
Tested with all screen-space AA methods in Forward+/Mobile, it works as expected.

Thanks!
